crux-lucene/src/crux/lucene.clj

```diff
@@ -0,0 +1,187 @@
+
+(defn- crux-doc->triples [crux-doc]
+  (->> (dissoc crux-doc :crux.db/id)
+       (mapcat (fn [[k v]]
+                 (for [v (if (coll? v) v [v])
+                       :when (string? v)]
+                   [(name k) v])))))
+
+(defrecord DocumentId [a v])
+
+(defn- ^Document triple->doc [[k ^String v]]
+  (doto (Document.)
+    ;; To search for triples by eid-a-v for deduping
+    (.add (StringField. "id", (eid->str (DocumentId. k v)), Field$Store/NO))
+    ;; The actual term, which will be tokenized
+    (.add (TextField. (name k), v, Field$Store/YES))
+    ;; Uses for wildcard searches
+    (.add (TextField. "_val", v, Field$Store/YES))
+    ;; The Attr (storage only, for temporal resolution)
+    (.add (StringField. "_attr", (name k), Field$Store/YES))))
+
+(defn- ^Term triple->term [[k ^String v]]
+  (Term. "id" (eid->str (DocumentId. k v))))
+
+(defn doc-count []
+  (let [{:keys [^Directory directory]} *node*
+        directory-reader (DirectoryReader/open directory)]
+    (.numDocs directory-reader)))
+
+(defn write-docs! [^IndexWriter index-writer docs]
+  (doseq [d docs t (crux-doc->triples d)]
+    (.updateDocument index-writer (triple->term t) (triple->doc t))))
+
+(defn evict! [indexer, node, eids]
+  (let [{:keys [^Directory directory ^Analyzer analyzer]} node
+        attrs-id->attr (->> (db/read-index-meta indexer :crux/attribute-stats)
+                            keys
+                            (map #(vector (mem/buffer->hex (cc/->id-buffer %)) %))
+                            (into {}))]
+    (with-open [index-snapshot (db/open-index-snapshot indexer)
+                index-writer (IndexWriter. directory, (IndexWriterConfig. analyzer))]
+      (let [qs (for [[a v] (db/exclusive-avs indexer eids)
+                     :let [a (attrs-id->attr (mem/buffer->hex a))
+                           v (db/decode-value index-snapshot v)]
+                     :when (not= :crux.db/id a)]
+                 (TermQuery. (Term. "id" (eid->str (DocumentId. (name a) v)))))]
+        (.deleteDocuments index-writer ^"[Lorg.apache.lucene.search.Query;" (into-array Query qs))))))
+
+(defn search [node, k, v]
+  (let [{:keys [^Directory directory ^Analyzer analyzer]} node
+        directory-reader (DirectoryReader/open directory)
+        index-searcher (IndexSearcher. directory-reader)
+        qp (if k
+             (QueryParser. (name k) analyzer)
+             (QueryParser. "_val" analyzer))
+        b (doto (BooleanQuery$Builder.)
+            (.add (.parse qp v) BooleanClause$Occur/MUST))
+        q (.build b)
+        q (FunctionScoreQuery. q (DoubleValuesSource/fromQuery q))
+        score-docs (.-scoreDocs (.search index-searcher q 1000))]
+
+    (when (seq score-docs)
+      (log/debug (.explain index-searcher q (.-doc ^ScoreDoc (first score-docs)))))
+
+    (cio/->cursor (fn []
+                    (.close directory-reader))
+                  (map (fn [^ScoreDoc d] (vector (.doc index-searcher (.-doc d))
```

Comment:
indentation here
```clojure
(map (fn [...]
       ...)
     score-docs)
```

otherwise it puts the onus on the reader to determine whether `score-docs` is part of the `fn` or not, by counting close parens

https://juxt.slack.com/archives/G018EA25K71/p1600275786012100